accounting/invoice/models.py

```diff
@@ -20,3 +20,387 @@
+        help_text=_("The unique invoice number. "
+                    "Defaults to yyyy-mm for the current year and month."))
+    date = models.DateTimeField(
+        default=timezone.now,
+        help_text=_("The date this invoice was created and sent for billing purposes. "
+                    "Defaults to right now, but can be changed."))
+    billing_start_date = models.DateTimeField(
+        default=utils.get_first_day_past_month,
+        help_text=_("The first date for which the line items in this invoice were provided. "
+                    "Defaults to the first day of the past month."))
+    billing_end_date = models.DateTimeField(
+        default=utils.get_last_day_past_month,
+        help_text=_("The last date for which the line items in this invoice were provided. "
+                    "Defaults to the last day of the past month."))
+    due_date = models.DateTimeField(
+        default=get_day_with_offset,
+        help_text=_("When this invoice should be paid."))
+    provider = models.ForeignKey(
+        Account, on_delete=models.CASCADE, related_name='provider_invoices',
+        help_text=_("The invoicing service/product provider."))
+    client = models.ForeignKey(
+        Account, on_delete=models.CASCADE, related_name='client_invoices',
+        help_text=_("The client being invoiced for services/products."))
+    paid = models.BooleanField(
+        default=False,
+        help_text=_("Whether this invoice has been paid by the client yet or not."))
+    extra_text = models.TextField(
+        blank=True, null=True,
+        help_text=_("Any arbitrary extra text that the provider would like to display on their invoice. "
+                    "Each template should have a designated location to place this extra text."))
+    signature = models.ImageField(
+        blank=True, null=True,
+        help_text=_("The provider's signature."))
+    template = models.CharField(
+        max_length=80, choices=InvoiceTemplate.choices, default=InvoiceTemplate.Default,
+        help_text=_("The template to use to generate this invoice."))
+    history = HistoricalRecords()
+
+    # TODO: The following 2 fields are temporary while we don't have a UI besides the Django admin.
+    # TODO: The front-end will eventually call endpoints to do what these fields allow us to do.
+    auto_download_jira_worklogs_on_save = models.BooleanField(  # pylint: disable=invalid-name
+        default=False,
+        help_text=_("Whether this invoice should automatically download JIRA worklogs "
+                    "for the provider to fill up the invoice."))
+    auto_create_pdf_on_save = models.BooleanField(
+        default=False,
+        help_text=_("Whether this invoice should be converted to a PDF on save."))
+    auto_upload_google_drive_on_save = models.BooleanField(  # pylint: disable=invalid-name
+        default=False,
+        help_text=_("Whether this invoice should be uploaded to Google Drive when converted to a PDF on save."))
```

Comment:
It is temporary, yeah. It relies on `auto_create_pdf_on_save` being set to have any effect: we can't upload a file whose path we have no clue about. The filename is generated very uniquely (concat 2 UUIDs) on the server, and has to be really unique for security reasons, because the media endpoint is public in Django. So only at run-time will the system know that this is the correct invoice to upload. (Aside: technically one can loop through the media files to find it based off of the filename with UUIDs cut off, but I don't think that's a good implementation at all).

Maybe there's another way to handle this securely but I'm not aware of it, or it might be a lot of extra work. For example, how would an admin see that this or that signature was uploaded, without being able to view it through some public endpoint, like `/media`?

Reply:
Why do you currently store the PDF in the media directory, rather than in a temporary file (for example using https://docs.python.org/3.6/library/tempfile.html )? Is it to allow to download that file, eventually, without requiring the admin to use Google Drive? If so you will need to store the path to the file eventually no?

Btw, is there a way at all currently for the admin to download the generated PDF directly from the django admin, ie a link to its location on Google Drive?

Reply:
The library I'm using to generate PDFs doesn't use `tempfile` so that's something we'd have to upstream if we wanted it (and I don't think it'd be to `pdfkit`, it might be to the actual `wkhtmltopdf` that does the outputting). All we can do right now is specify the output path. So it would have to inevitably go somewhere, where we can retrieve it, and the most common sense place would be `/media`, except that it's public.

But I see what you're saying: we can't then retrieve it eventually for download unless we have the path. So I think the better way would be to upload these to S3 or similar. But since we're already doing this for Google Drive, we can just make GDrive what S3 would be for us, and have it be available for download through our integration there. Of course, that's for when we get to that point with a real admin UI. For now just pick it up from GDrive directly.

Reply:
@UmanShahzad 

> The library I'm using to generate PDFs doesn't use tempfile so that's something we'd have to upstream if we wanted it (and I don't think it'd be to pdfkit, it might be to the actual wkhtmltopdf that does the outputting). All we can do right now is specify the output path. So it would have to inevitably go somewhere, where we can retrieve it, and the most common sense place would be /media, except that it's public.

Then can't you pass on `tempfile.mkdtemp()` as the destination directory?

> But I see what you're saying: we can't then retrieve it eventually for download unless we have the path. So I think the better way would be to upload these to S3 or similar. But since we're already doing this for Google Drive, we can just make GDrive what S3 would be for us, and have it be available for download through our integration there. Of course, that's for when we get to that point with a real admin UI. For now just pick it up from GDrive directly.

Yes, just using Google Drive for storing/downloading the file seem completely fine.

> nope, there's no way yet through the Django admin directly. I don't think a link is needed, right? It's always going to go in the same place (okay, changes for the month, of course), and I believe GDrive can update live.

I think that was part of the requirements I had listed for what to display in the admin? We can of course go locate it manually in Google Drive, but having to do this for each invoice is going to be repetitive and take time, so it would be really useful to have a direct link to the file in Google Drive, like we have currently in the spreadsheets Alicia builds by hand.

Reply:
@antoviaque 

> Then can't you pass on tempfile.mkdtemp() as the destination directory, or tempfile.mkstemp()[1] as the destination file?

Those are lower-level functions that require us to manually clean it up ourselves. That wouldn't be different than what we're already doing.

Either way, I'm starting to lose sight of what the purpose of `tempfile` would be for us. If we're looking to download the PDF file, we can't use it anyway.

> I think that was part of the requirements I had listed for what to display in the admin? We can of course go locate it manually in Google Drive, but having to do this for each invoice is going to be repetitive and take time, so it would be really useful to have a direct link to the file in Google Drive, like we have currently in the spreadsheets Alicia builds by hand.

So let's decide to use a `pdf_path` URL field to store its location, whether that's local or on GDrive. If GDrive integration is enabled, it can be uploaded there by default, and the link would be stored in the field so it can be copy/pasted. Otherwise it'll be a filesystem URL. In either case, we can use the URL in the future to download the PDF.

Reply:
@UmanShahzad Sounds good for the approach - thanks for adding the link!

PS: To illustrate on a point from our discussion on the epic - on changes like this it's fine to put them in a ticket for the following sprint, I don't necessarily expect you to address all the product related points in the current PR. Just so you can understand a bit more how to work with me on these points, without ending up with an inflated sprint scope every time I review your work : ) Feel free to ask also, if you're not sure of the importance/urgency of a point, I'm happy to spend time agreeing on this to make sure we have a flow that is comfortable for both of us.

Reply:
@antoviaque Yeah, I think that's the issue: it's natural that we'll have a lot to change in some cases, but instead of making new tickets to address these problems which may be non-trivial, I'm trying to stretch my sprint to fit them all in right now, which is impractical and just overall problematic.

So from now, after @itsjeyd approves from his side, you can do the review as you've been doing, and anything that I think I can fit in reasonably into the sprint, I will address, and anything that I can't, I'll make a dedicated ticket for the following sprint, given higher priority over other tickets, with a reference to the comment and some context.

Does that approach sound good to you guys, @itsjeyd and @antoviaque ?

Reply:
And thanks for discussing all of this -- I'm glad we're coming to something that'll work well for the all of us.

Reply:
@UmanShahzad That approach sounds good to me!